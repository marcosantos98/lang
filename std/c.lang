extern "c" puts :: int(str char*)
extern "c" putchar :: int(d int)
extern "c" exit :: int(code int)
extern "c" printf :: int(str char*, args ...)
