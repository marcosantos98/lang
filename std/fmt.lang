import "std:c"

putd :: void(d int) {
    if d > 9 {
	var a = d / 10
	d -= 10 * a
	putd(a)
    }
    putchar(48 + d)
}
