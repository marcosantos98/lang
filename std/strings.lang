import "std:c"

struct StringBuilder {
    data ^char,
    len int,
    cap int,
    a string
}

sb_new :: StringBuilder() {
    var data = malloc(1024)
    memset(data, 0, 1024)
    return StringBuilder{data as ^char, 0, 1024}
}

sb_write_str :: void(sb ^StringBuilder, append string) {
    var new_len = sb.len + len(append)
    if new_len > sb.cap {
	sb.cap = sb.cap * 2
	sb.data = realloc(sb.data, sb.cap) as ^char
    }
    var ptr: ^char = sb.data as ^char
    memcpy(ptr + sb.len, append.data, len(append))
    sb.len = new_len
}

sb_print :: void(sb StringBuilder) {
    printf(c("%.*s"), sb.len, sb.data)
}

sb_to_cstr :: ^char(sb StringBuilder) {
    return sb.data as ^char
}

sb_destroy :: void(sb StringBuilder) {
    free(sb.data)
}
