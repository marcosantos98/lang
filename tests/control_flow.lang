import "std:c"
import "std:fmt"

main :: void() {

    var i = 0
    if i < 10 {
	puts(xx to_cstring("Hello, World"))
    }

    var k = 10
    if k > 10 {
	puts(xx to_cstring("K is greater than 10"))
    } else if k < 10 {
	puts(xx to_cstring("K is less than 10"))
    } else {
	printf(xx to_cstring("K is %d\n"), k)
    }

    while i < 2 {
	puts(xx to_cstring("I is less than 2"))
	i += 1
    }

    for var j = 0; j < 10; j += 1 {
	putd(j)
	puts(xx to_cstring(""))
    }

    var l = 0
    while true {
	puts(xx to_cstring("Should print 4 times"))
	if l < 3 {
	    l += 1
	} else {
	    break
	}
    }
}
