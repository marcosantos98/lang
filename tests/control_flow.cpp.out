/* -- Builtin types -- */
typedef const char* cstr;
/* -- ------------- -- */

extern "C" int puts(cstr);
extern "C" int putchar(int);
extern "C" int exit(int);
void putd(int);
void ___entry___();

void putd(int d){
if (d>9){
int a = d/10;
d -= 10*a;
putd(a);
}
putchar(48+d);
}
void ___entry___(){
int i = 0;
if (i<10){
puts("Hello, World");
}
int k = 100;
if (k>10){
puts("K is greater than 10");
}
 else if (k<10){
puts("K is less than 10");
}
 else {
puts("K is 10");
}
while (i<2){
puts("I is less than 2");
i += 1;
}
for(int j = 0;j<10;j += 1){
puts("TODO: add print number or printf support");
}
int l = 0;
while (true){
puts("Should print 4 times");
if (l<3){
l += 1;
}
 else {
break;
}
}
}
int main(int argc, char** argv) {___entry___();}
